r/R/dplyr-funcs-datetime.R

```diff
@@ -233,6 +236,68 @@ register_bindings_datetime <- function() {
                                        tz = "UTC") {
     call_binding("make_datetime", year, month, day, hour, min, sec, tz)
   })
+  register_binding("difftime", function(time1,
+                                        time2,
+                                        tz,
+                                        units = "secs") {
+    if (units != "secs") {
+      abort("`difftime()` with units other than `secs` not supported in Arrow")
+    }
+
+    if (!missing(tz)) {
+      warn("`tz` argument is not supported in Arrow, so it will be ignored")
+    }
+
+    # cast to timestamp if time1 and time2 are not dates or timpestamp expressions
+    # (the subtraction of which would output a `duration`)
+    if (!(inherits(time1, "Expression") &&
+          time1$type_id() %in% Type[c("TIMESTAMP", "DATE32", "DATE64")])) {
+      time1 <- build_expr("cast", time1, options = cast_options(to_type = timestamp(timezone = "UTC")))
+    }
+
+    if (!(inherits(time2, "Expression") &&
+          time2$type_id() %in% Type[c("TIMESTAMP", "DATE32", "DATE64")])) {
+      time2 <- build_expr("cast", time2, options = cast_options(to_type = timestamp(timezone = "UTC")))
+    }
```

Comment:
I think we might have talked about this (though it might have been in a different PR and not this one — I've looked through the comments and haven't found it). 

Could we use `!call_binding("is.instant", time1)` in the `if` conditions here? The expression bits are the same, and `build_expr` should take care of the converting the R types to Arrow types. 

Or do we need to do something totally different if we get R objects (e.g. always skip to line 263 below)? If that's the case, maybe we should do something like `if (!inherits(time1, "Expression") && !call.binding(is.instant, time1)) {` to make that a bit clearer?

Reply:
This is the final line in that binding (row 264):

```r
build_expr("cast", time1 - time2, options = cast_options(to_type = duration("s")))
```

if we use `!call_binging("is.instant", time1)` it means that when `time1` is a POSIXct object it will jump straight to it. We get an error because of the subtraction operation `-`. 

```r
via_table <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(.input = arrow_table(tbl))))` threw an unexpected warning.
Message: Incompatible methods ("-.POSIXt", "Ops.Expression") for "-"
Class:   simpleWarning/warning/condition
```  

I guess we could build a subtraction expression, but that feels more complicated to me - to big a trade-off for using  `is.instant`.

Reply:
For what it's worth the behavior you've run into here should probably have a jira + be cleaned up. But I'll leave it up to you as to how you want to work around that here (with `is.instant` + other catches or the approach you have already)

Reply:
on a second thought, the `"is.instant"` binding + `"subtract_checked"` Expression isn't that clunky and a bit cleaner, so we could go with that:
```r
    if (!call_binding("is.instant", time1)) {
      time1 <- build_expr("cast", time1, options = cast_options(to_type = timestamp(timezone = "UTC")))
    }

    if (!call_binding("is.instant", time2)) {
      time2 <- build_expr("cast", time2, options = cast_options(to_type = timestamp(timezone = "UTC")))
    }

    # we need to do this instead of `time1 - time2` to prevent complaints when
    # we try to subtract an R object from an Expression
    subtract_output <- build_expr("subtract_checked", time1, time2)
    build_expr("cast", subtract_output, options = cast_options(to_type = duration("s")))
```

Reply:
`build_expr("-", time1, time2)` works too

Reply:
This is the behaviour that seemd off:

```
via_table <- rlang::eval_tidy(expr, rlang::new_data_mask(rlang::env(.input = arrow_table(tbl))))` threw an unexpected warning.
Message: Incompatible methods ("-.POSIXt", "Ops.Expression") for "-"
Class:   simpleWarning/warning/condition
```

But maybe that's actually a result of setting up the expressions wrong. You should check the tests + code for how we define these generics and see if one is missing such that someone might run into this or if it's that the code that produced that error was an error in the PR instead and that's a setup that would never happen for a user.

Reply:
I do think that is a bit PR specific, as it's hard to envisage a situation in which a user will be faced with subtracting an R object from an Expression.